app/controllers/replies_controller.rb

```diff
@@ -297,19 +300,147 @@ def require_edit_permission
     redirect_to post_path(@reply.post)
   end
 
-  def preview(written)
-    @written = written
-    @post = @written.post
+  def get_multi_replies
+    @multi_replies_json = JSON.parse(params.fetch(:multi_replies_json, "[]"))
+    @multi_replies = @multi_replies_json.map do |reply_json|
+      Reply.new(permitted_params(ActionController::Parameters.new({ reply: reply_json })))
+        .tap { |r| r.user = current_user }
+    end
+  end
+
+  def preview_reply(reply)
+    # Previewing a specific reply
+    @post = reply.post
+    @written = reply
     @written.user = current_user unless @written.user
     @audits = @written.id.present? ? { @written.id => @written.audits.count } : {}
+    @reply = @written # So that editor_setup shows the correct characters
+    preview_replies
+  end
+
+  def add_to_multi_reply(reply, reply_params)
+    # Adding a new reply to a multi reply
+    post_id = params[:reply][:post_id]
+    ReplyDraft.draft_for(post_id, current_user.id).try(:destroy)
+
+    # Save NPC
+    if reply.character&.new_record?
+      if reply.character.save
+        flash[:success] = "Your new NPC has been persisted!"
+        params[:reply][:character_id] = reply.character.id
+        reply_params[:character_id] = reply.character.id
+        reply.character_id = reply.character.id
+      else
+        flash[:error] = "There was a problem persisting your new NPC."
+      end
+    end
+
+    # Set up editor
+    @post = reply.post
+    empty_reply_hash = permitted_params.permit(:character_id, :icon_id, :character_alias_id)
+    @empty_written = @post.build_new_reply_for(current_user, empty_reply_hash)
+    @empty_written.editor_mode ||= params[:editor_mode] || current_user.default_editor
+    @reply = @empty_written # So that editor_setup shows the correct characters
+    @audits = {}
+
+    # Add reply to list of multi replies
+    reply.user = current_user unless reply.user
+    @multi_replies << reply
+    reply_params[:id] = reply.id if reply.id.present?
+    @multi_replies_json << reply_params
+
+    preview_replies
+  end
 
+  def preview_replies
     @page_title = @post.subject
 
-    @reply = @written # So that editor_setup shows the correct characters
     editor_setup
     render :preview
   end
 
+  def post_replies(new_reply: nil)
+    @multi_replies << new_reply if new_reply.present?
+
+    if @multi_replies_json.present? && (@reply = Reply.find_by_id(@multi_replies_json.first["id"]))
+      # The first reply of the multi replies has an ID, that means I'm editing rather than posting a new one
+      original_reply_params = permitted_params(ActionController::Parameters.new({ reply: @reply.attributes }))
```

Comment:
Use `@reply.dup`? Then you can pass a `Reply` object around instead of doing a `permitted_params(... @reply.attributes ...)` mutation